File: api/api-lambdas/putevents/putevents.js
const AWS = require('aws-sdk')
const { createMetricsLogger, Unit } = require("aws-embedded-metrics");
//var moment = require('moment-timezone');

var personalizeevents = new AWS.PersonalizeEvents();
//var dynamoClient = new AWS.DynamoDB.DocumentClient();

console.log('Loading function');

function getTimeZoneOffset(date, timeZone) {

  // Abuse the Intl API to get a local ISO 8601 string for a given time zone.
  let iso = date.toLocaleString('en-CA', { timeZone, hour12: false }).replace(', ', 'T');

  // Include the milliseconds from the original timestamp
  iso += '.' + date.getMilliseconds().toString().padStart(3, '0');

  // Lie to the Date object constructor that it's a UTC time.
  const lie = new Date(iso + 'Z');

  // Return the difference in timestamps, as minutes
  // Positive values are West of GMT, opposite of ISO 8601
  // this matches the output of `Date.getTimeZoneOffset`
  return -(lie - date) / 60 / 1000;
}

exports.handler = (event, context, callback) => {
    console.log(JSON.stringify(event, null, 2));
    
    event.Records.forEach(function(record) {
        // Kinesis data is base64 encoded so decode here
        const metrics = createMetricsLogger();
        metrics.putDimensions({ Type: "PutEvents" });

        var payload = Buffer.from(record.kinesis.data, 'base64').toString('ascii');
        //console.debug('Decoded payload:', payload);
        payload = JSON.parse(payload);

        if (payload.sp_event_id != undefined){
            metrics.setProperty("EventID", payload.sp_event_id);
        }

        if (payload.sp_derived_tstamp === undefined || payload.sp_derived_tstamp.trim().length === 0){
            console.debug("Skipping event: not valid sp_derived_tstamp")
            metrics.putMetric("EventStatus", 0);
            metrics.flush();
            return context.successful;
        }
        
        var eventDate = new Date(payload.sp_derived_tstamp);

        if (payload.sp_app_id === undefined || payload.sp_app_id != process.env.FilterAppId){
            console.debug("Skipping event: not valid sp_app_id")
            metrics.putMetric("EventStatus", 0);
            metrics.flush();
            return context.successful;
        }

        if (payload.content_contentId === undefined || payload.content_contentId.trim().length === 0){
            console.debug("Skipping event: not valid content_contentId")
            metrics.putMetric("EventStatus", 0);
            metrics.flush();
            return context.successful;
        }
        
        if (payload.page_type === undefined || payload.page_type != "article"){
            console.debug("Skipping event: not valid page_type")
            metrics.putMetric("EventStatus", 0);
            metrics.flush();
            return context.successful;
        }
        
        if (payload.sp_event_name === undefined || payload.sp_event_name != "page_view"){
            console.debug("Skipping event: not valid sp_event_name")
            metrics.putMetric("EventStatus", 0);
            metrics.flush();
            return context.successful;
        }

        if (payload.sp_domain_sessionid === undefined){
            console.debug("Skipping event: missing sp_domain_sessionid")
            metrics.putMetric("EventStatus", 0);
            metrics.flush();
            return context.successful;
        }

        if ((payload.sp_user_id === undefined || payload.sp_user_id.trim().length === 0) && (payload.sp_domain_userid === undefined || payload.sp_domain_userid.trim().length === 0)){
            console.debug("Skipping event: missing sp_user_id and sp_domain_userid")
            metrics.putMetric("EventStatus", 0);
            metrics.flush();
            return context.successful;
        }
        
        var putEventsParams= {
            'sessionId': payload.sp_domain_sessionid,
            'trackingId': process.env.TRACKING_ID,
            'userId': payload.sp_user_id ? payload.sp_user_id : payload.sp_domain_userid,
            eventList: [
                {
                  'eventId': payload.sp_evenet_id,
                  'eventType': payload.sp_event_name, 
                  'itemId': payload.content_contentId,
                  'sentAt': eventDate,
                 
                  'properties': {
                      'visitor_type': payload.visitor_type,
                      'visitor_countryCode': payload.visitor_countryCode,
                      'device_detector_visitorPlatform': payload.device_detector_visitorPlatform,
                      'device_detector_brandName': payload.device_detector_brandName,
                      'device_detector_browserFamily': payload.device_detector_browserFamily,
                      //'ContentText': null,
                      //'Category': null,
                      //'WordCount': null
                  }, 
                },
            ]
        }
        
        if (payload.page_rid != undefined){
            putEventsParams.eventList[0].recommendationId = payload.page_rid
        }
      
        console.log("THIS IS THE OBJECT = " + JSON.stringify(putEventsParams,null,3))
        console.log("THIS IS THE SOURCE OBJECT = " + JSON.stringify(payload,null,3))

        personalizeevents.putEvents(putEventsParams, function (err, data) {
          if (err) {
                console.log(err, err.stack); // an error occurred
                metrics.putMetric("EventStatus", -1);
          }
          else{ 
                metrics.putMetric("EventStatus", 1);
                metrics.flush();
                console.log("Success: " + JSON.stringify(data, null, 2)) 
                //console.log(data);           // successful response
                putEventsParams['eventList'][0]['sentAt']=putEventsParams['eventList'][0]['sentAt'].toTimeString();
                const putEventsErrResponse = {
                    statusCode: 500,
                    body: JSON.stringify(err),
                };
                callback(null, putEventsErrResponse);
                const response = {
                    statusCode: 200,
                    body: JSON.stringify(putEventsParams),
                };
                callback(null, response);
          }
        });
    
    });
};

File: api/api-lambdas/putEvents/putEvents.py
import time
import json
import os
import base64

from boto3 import client

from botocore import config as boto_config
from botocore.exceptions import ClientError
from aws_embedded_metrics import metric_scope

config = boto_config.Config(
    connect_timeout=1, read_timeout=1,
    retries={'max_attempts': 2})

enviroment            = os.environ.get('Environment')
resources_prefix      = os.environ.get('ResourcesPrefix')
filter_app_id         = os.environ.get('FilterAppId')
personalize_cli       = client('personalize-events', config=config)
dynamo_client         = client('dynamodb', config=config)
settings_dynamo_table =  f"{resources_prefix}-{enviroment}-api-settings"


def _get_dynamo_settings(db, dynamo_table):
    try:
        personalize_list = []
        scan_kwargs = {
            'TableName': dynamo_table,
            'FilterExpression': "#status = :active ",
            'ExpressionAttributeNames': {
                "#status": "status",
            },
            'ExpressionAttributeValues': {
                ":active": {
                    'S': "active"
                }
            }
        }

        done = False
        start_key = None
        while not done:
            if start_key:
                scan_kwargs['ExclusiveStartKey'] = start_key
            response = db.scan(**scan_kwargs)
            personalize_list.extend(response.get('Items', []))
            start_key = response.get('LastEvaluatedKey', None)
            done = start_key is None

        return personalize_list

    except ClientError as e:
        print(f"Key Error: {e}")


@metric_scope
def handler(event, context, metrics):
    metrics.set_property("LambdaRequestId", context.aws_request_id)
    metrics.put_dimensions({"Type": "PutEvents"})
    settings = _get_dynamo_settings(dynamo_client, settings_dynamo_table)

    if len(settings) == 0:
        return {'statusCode': '500', 'body': json.dumps("No active trackers found in dynamo settings table")}

    success_events = 0
    fail_events = 0
    skip_events = 0

    for record in event['Records']:
        #Kinesis data is base64 encoded so decode here
        payload_str=base64.b64decode(record["kinesis"]["data"])
        payload = json.loads(payload_str)
        #print("Decoded payload: " + str(payload.get('sp_event_id')))

        if (payload.get('sp_event_id') is None):
            printf(f"Skipping event: invalid sp_event_id {payload.get('sp_event_id')}")
            skip_events += 1;
            continue

        if (payload.get('sp_derived_tstamp') is None) or len(payload.get('sp_derived_tstamp').strip()) == 0:
            print(f"Skipping event: invalid sp_derived_tstamp: {payload.get('sp_derived_tstamp')}")
            skip_events += 1;
            continue

        if (payload.get('sp_app_id') is None) or payload.get('sp_app_id') != filter_app_id:
            print(f"Skipping event: invalid sp_app_id: {payload.get('sp_app_id')}")
            skip_events += 1;
            continue

        if (payload.get('content_contentId') is None) or len(payload.get('content_contentId').strip()) == 0:
            print(f"Skipping event: invalid content_contentId: {payload.get('content_contentId')}")
            skip_events += 1;
            continue

        if (payload.get('page_type') is None) or payload.get('page_type') != "article":
            print(f"Skipping event: invalid page_type: {payload.get('page_type')}")
            skip_events += 1;
            continue

        if (payload.get('sp_event_name') is None) or payload.get('sp_event_name') != "page_view":
            print(f"Skipping event: invalid sp_event_name: {payload.get('sp_event_name')}")
            skip_events += 1;
            continue

        if (payload.get('sp_domain_sessionid') is None):
            print(f"Skipping event: invalid sp_domain_sessionid {payload.get('sp_domain_sessionid')}")
            skip_events += 1;
            continue

        if (payload.get('sp_user_id') is None or (len(payload.get('sp_user_id').strip()) == 0)) and (payload.get('sp_domain_userid') is None or (len(payload.get('sp_domain_userid').strip()) == 0)):
            print(f"Skipping event: invalid content_contentId: {payload.get('content_contentId')}")
            skip_events += 1;
            continue

        if ((payload.get('sp_user_id') is not None) and len(payload.get('sp_user_id').strip()) > 0):
            user_id = payload.get('sp_user_id')
        if ((payload.get('sp_domain_userid') is not None) or len(payload.get('sp_domain_userid').strip()) > 1):
            user_id = payload.get('sp_domain_userid')

        try:
            timestamp = int(round(time.mktime(time.strptime(payload.get('sp_derived_tstamp')[:-5], '%Y-%m-%dT%H:%M:%S'))))
        except:
            timestamp = int(round(time.time()))

        putEventsParams = {
            'sessionId': payload.get('sp_domain_sessionid'),
            'userId': user_id,
            'eventList': [{
                'eventId': payload.get('sp_event_id'),
                'eventType':  payload.get('sp_event_name'),
                'itemId':  payload.get('content_contentId'),
                'sentAt':  timestamp,
                'properties':  json.dumps({
                    'visitor_type': payload.get('visitor_type'),
                    'visitor_countryCode': payload.get('visitor_countryCode'),
                    'device_detector_visitorPlatform': payload.get('device_detector_visitorPlatform'),
                    'device_detector_brandName': payload.get('device_detector_brandName'),
                    'device_detector_browserFamily': payload.get('device_detector_browserFamily'),
                })
            }]
        }

        if (payload.get('page_rid') is not None):
            putEventsParams['eventList'][0]['recommendationId'] = payload.get('page_rid')

        status_code = "200"
        status_body = json.dumps("Success")

        for tracker in settings:
            print(f"Put event to tracker {tracker.get('eventTrackerId').get('S')}")
            putEventsParams['trackingId'] = tracker.get('eventTrackerId').get('S')
            print("This is the input object: " + str(putEventsParams))
            try:
                personalize_cli.put_events(**putEventsParams)
            except ClientError as e:
                status_code = "500"
                status_body = f"Personalize Client Error: {e}"
                print(f"Personalize Client Error: {e}")
                fail_events += 1
            else:
                success_events += 1

    metrics.put_metric("SuccessEvents", success_events, "None")
    metrics.put_metric("FailEvents", fail_events, "None")
    metrics.put_metric("SkipEvents", skip_events, "None")
    return {'statusCode': status_code, 'body': status_body}



